### PR TITLE
Allow inverting non-unique relationships

### LINKED ISSUE #3006: RequiredAttribute on navigation should try to set dependent side
`Required` attribute on navigation property defines the null-ability of foreignkey. Hence it is logical to have `Required` appear on the navigation on dependent side (i.e. dependent to principal navigation). If we encounter `Required` on principal to dependent navigation, we should try inverting relationship and set required.
In case of required being on both navigations, do not invert, set required and log warning.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Discover fk properties if principal end is configured by convention (not ambiguous)
Make RequiredAttribute configure the principal end when specified on only one navigation and principal end wasn't configured by convention

Fixes #3006

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore/Metadata/Conventions/Internal/NavigationAttributeNavigationConvention.cs [hunk 1]
```diff
@@ -35,5 +36,12 @@ public virtual InternalRelationshipBuilder Apply(InternalRelationshipBuilder rel
         }
 
         public abstract InternalRelationshipBuilder Apply([NotNull] InternalRelationshipBuilder relationshipBuilder, [NotNull] Navigation navigation, [NotNull] TAttribute attribute);
+
+        protected static IEnumerable<TCustomerAttribute> GetAttributes<TCustomerAttribute>(EntityType entityType, string propertyName)
+            where TCustomerAttribute : Attribute
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Metadata/Conventions/Internal/NavigationAttributeNavigationConvention.cs]: `TCustomAttribute` better?

Comment (AndriySvyryd) [inline src/Microsoft.EntityFrameworkCore/Metadata/Conventions/Internal/NavigationAttributeNavigationConvention.cs]: haha, nice spot

File: src/Microsoft.EntityFrameworkCore/Metadata/Internal/InternalRelationshipBuilder.cs [hunk 2]
```diff
@@ -261,56 +208,49 @@ public class InternalRelationshipBuilder : InternalMetadataItemBuilder<ForeignKe
...
                     return false;
                 }
             }
 
-            var conflictingNavigation = navigationName == null
-                ? null
-                : entityTypeBuilder.Metadata.FindNavigation(navigationName);
-            if (conflictingNavigation != null)
+            if (navigationName == null)
+            {
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Metadata/Internal/InternalRelationshipBuilder.cs]: can be merged as else in previous if.

File: src/Microsoft.EntityFrameworkCore/Metadata/Internal/InternalRelationshipBuilder.cs [hunk 3]
```diff
@@ -770,7 +715,12 @@ private bool CanInvert(ConfigurationSource configurationSource)
             }
 
             var resetIsRequired = false;
-            if (!CanSetRequiredOnProperties(properties, Metadata.IsRequired, configurationSource, shouldThrow: false))
+            if (!CanSetRequiredOnProperties(
+                properties,
+                Metadata.IsRequired,
+                Metadata.DeclaringEntityType,
+                configurationSource,
+                false))
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Metadata/Internal/InternalRelationshipBuilder.cs]: parameter name for Boolean.

File: src/Microsoft.EntityFrameworkCore/Metadata/Internal/InternalRelationshipBuilder.cs [hunk 4]
```diff
@@ -829,13 +780,18 @@ private bool CanInvert(ConfigurationSource configurationSource)
...
 
-            if (!configurationSource.Overrides(Metadata.GetIsRequiredConfigurationSource())
-                && !CanSetRequiredOnProperties(properties, Metadata.IsRequired, configurationSource, shouldThrow: false))
+            if (!configurationSource.Value.Overrides(Metadata.GetIsRequiredConfigurationSource())
+                && !CanSetRequiredOnProperties(
+                    properties,
+                    Metadata.IsRequired,
+                    Metadata.DeclaringEntityType,
+                    configurationSource,
+                    false))
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Metadata/Internal/InternalRelationshipBuilder.cs]: parameter name

**Comment by smitpatel** [comment]
:shipit:

**Comment by AndriySvyryd** [comment]
@smitpatel I had to significantly change how existing relationships are matched in `ReplaceForeignKey` to make all the tests pass. Could you review the new commit?

**Comment by smitpatel** [comment]
:shipit:

**Pull request merged**